src/tokens/TokenImport.tsx

```diff
@@ -27,57 +30,149 @@ import variables from 'src/styles/variables'
+  decimals: number
+  name: string
+  balance: bigint
+}
+
 export default function TokenImportScreen(_: Props) {
   const { t } = useTranslation()
   const dispatch = useDispatch()
 
-  const [isCompleteAddress, setIsCompleteAddress] = useState(false)
   const [tokenAddress, setTokenAddress] = useState('')
   const [tokenSymbol, setTokenSymbol] = useState('')
   const [error, setError] = useState<string | null>(null)
   const [networkId] = useState(networkConfig.defaultNetworkId)
+
+  const walletAddress = useSelector(walletAddressSelector)
   const supportedTokens = useSelector((state) => tokensByIdSelector(state, [networkId]))
 
-  const validateAddress = (address: string) => {
-    if (!address) return
-
-    if (isAddress(address)) {
-      setIsCompleteAddress(true)
-      // TODO(RET-891): if already imported, set state as AddressState.AlreadyImported
-      const tokenId = getTokenId(networkId, address.toLowerCase())
-      if (supportedTokens[tokenId]) {
-        setError(t('tokenImport.error.alreadySupported'))
-      } else {
-        setError(null)
-      }
-    } else {
+  const validateAddress = (address: string): address is Address => {
+    if (!address) return false
+
+    if (!isAddress(address)) {
       setError(t('tokenImport.error.invalidToken'))
+      return false
+    }
+
+    // TODO(RET-891): if already imported, set state as AddressState.AlreadyImported
+    const tokenId = getTokenId(networkId, address.toLowerCase())
+    if (supportedTokens[tokenId]) {
+      setError(t('tokenImport.error.alreadySupported'))
+      ValoraAnalytics.track(AssetsEvents.import_token_already_supported, {
+        networkId,
+        tokenId,
+        tokenAddress,
+      })
+      return false
     }
-  }
 
-  const handleAddressFocus = () => {
-    setIsCompleteAddress(false)
     setError(null)
+    return true
+  }
+
+  const fetchTokenDetails = async (
+    tokenAddress: Address,
+    walletAddress: Address
+  ): Promise<TokenDetails> => {
+    const contract = getContract({
+      abi: erc20.abi,
+      address: tokenAddress,
+      publicClient: publicClient[networkIdToNetwork[networkId]],
+    })
+
+    const [symbol, decimals, name, balance] = await Promise.all([
+      contract.read.symbol(),
+      contract.read.decimals(),
+      contract.read.name(),
+      contract.read.balanceOf([walletAddress]),
+    ])
+
+    return { address: tokenAddress, symbol, decimals, name, balance }
   }
 
+  const validateContract = useAsyncCallback(
+    async (address: Address): Promise<TokenDetails> => {
+      if (!walletAddress || !isAddress(walletAddress)) {
+        // should never happen
+        throw new Error('No wallet address found when fetching token details')
+      }
+
+      const details = fetchTokenDetails(address, walletAddress)
+      const timeout = new Promise<never>((_, reject) =>
+        setTimeout(reject, FETCH_TOKEN_DETAILS_TIMEOUT_MS, new Error('Timeout'))
+      )
+
+      return Promise.race([timeout, details])
```

Comment:
promise.race doesn't cancel the in flight promises...i couldn't find an easy way to specify a timeout in the viem requests to ensure no unresolved promises when the timeout is hit :/ so i'm just noting this thought here without having a good suggestion of how to handle this here. one strawperson proposal i have is to implement no timeout (since i think there is a 60s default timeout with viem for RPC calls) and we can always measure the performance of this action with Sentry as a follow up if we are concerned. since this is not a routine user action, it seems acceptable if the request takes a long time (in comparison to a user with a very slow / bad connection that always hits the timeout can never import a token)

Reply:
`+1` to just rely on Viem API doing it's thing. they have a sophisticated [timeout handling](https://github.com/wevm/viem/blob/e74c9ab92a8b0635b7f99512d5276527421a2bf4/src/utils/rpc.ts#L92).

Reply:
Thanks, initially I could not find Viem's timeout settings. The default HTTP transport uses a timeout of 10s.

I'll remove the Promise.race and the custom timeout function.